storage/s3.go

```diff
@@ -382,6 +386,175 @@ func (s *S3) Get(
 	})
 }
 
+type SelectRecord struct {
+	Record json.RawMessage
+	Err    error
+}
+
+func (s *S3) selectStream(resp *s3.SelectObjectContentOutput, resultch chan<- *SelectRecord) {
+	defer resp.EventStream.Close()
+
+	reader, writer := io.Pipe()
+	defer reader.Close()
+	go func() {
+		defer writer.Close()
+		// TODO: wire ctx up to eventStream.StreamCloser.Close
+		for event := range resp.EventStream.Reader.Events() {
+			switch e := event.(type) {
+			case *s3.RecordsEvent:
+				// TODO: how should we handle errors here?!
+				writer.Write(e.Payload)
+			// How the heck are these different?
+			case *s3.StatsEvent:
+			case *s3.ProgressEvent:
+			}
+		}
+	}()
+
+	decoder := json.NewDecoder(reader)
+	// decoder.DontCopyRawMessage()
+	for {
+		var record json.RawMessage
+		err := decoder.Decode(&record)
+		if err != nil {
+			if err != io.EOF {
+				resultch <- &SelectRecord{
+					Err: err,
+				}
+			}
+			return
+		}
+		resultch <- &SelectRecord{
+			Record: record,
+		}
+	}
+	err := resp.EventStream.Err()
+	if err != nil {
+		resultch <- &SelectRecord{
+			Err: err,
+		}
+	}
+}
+
+func (s *S3) selectWorker(ctx context.Context, input s3.SelectObjectContentInput, objch <-chan *Object, resultch chan<- *SelectRecord) {
+	for object := range objch {
+		// TODO: reintroduce
+		if object.Type.IsDir() { // || errorpkg.IsCancelation(object.Err) {
```

Comment:
I can't `import "errorpkg"` because it causes an import loop. If you have a preferred way of resolving this, lmk.

Reply:
`errorpkg` is (implicitly) meant to be used at the command layer. Almost all of the storage errors bubble up to the callsite, handled there.